prow/genyaml/genyaml.go

```diff
@@ -0,0 +1,306 @@
+	"go/token"
+	yaml3 "gopkg.in/yaml.v3"
+	"reflect"
+	"strings"
+)
+
+var CommentMap map[string]map[string]Comment
+
+func init() {
+	const defaultPath = "prow/plugins/config.go"
+	CommentMap, _ = genDocMap(defaultPath)
+}
+
+// Comment is an abstract structure for storing parsed AST comments decorated with contextual information.
+type Comment struct {
+	// Type is the underlying type of the identifier associated with the comment.
+	Type string
+	// IsObj determines if the underlying type is a object type (e.g. struct) or primitive type (e.g. string).
+	IsObj bool
+	// Doc is a comment string parsed from the AST of a node.
+	Doc string
+}
+
+// astFrom takes a path to a Go file and returns the abstract syntax tree (AST) for that file.
+func astFrom(path string) (*doc.Package, error) {
+	fset := token.NewFileSet()
+	m := make(map[string]*ast.File)
+
+	f, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse file to AST from path: %s", path)
+	}
+
+	m[path] = f
+	apkg, _ := ast.NewPackage(fset, m, nil, nil)
+
+	astDoc := doc.New(apkg, "", 0)
+	if astDoc == nil {
+		return nil, fmt.Errorf("unable to parse AST documentation from path: %s", path)
+	}
+
+	return astDoc, nil
+}
+
+// fmtRawDoc formats/sanitizes a Go doc string removing TODOs, newlines, whitespace, and various other characters from the resultant string.
+func fmtRawDoc(rawDoc string) string {
+	var buffer bytes.Buffer
+	delPrevChar := func() {
+		if buffer.Len() > 0 {
+			buffer.Truncate(buffer.Len() - 1) // Delete the last " " or "\n".
+		}
+	}
+
+	// Ignore all lines after ---.
+	rawDoc = strings.Split(rawDoc, "---")[0]
+
+	for _, line := range strings.Split(rawDoc, "\n") {
+		line = strings.TrimRight(line, " ")
+		leading := strings.TrimLeft(line, " ")
+		switch {
+		case len(line) == 0: // Keep paragraphs
+			delPrevChar()
+			buffer.WriteString("\n\n")
+		case strings.HasPrefix(leading, "TODO"): // Ignore one line TODOs.
+		case strings.HasPrefix(leading, "+"): // Ignore instructions to the generators.
+		default:
+			if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
+				delPrevChar()
+				line = "\n" + line + "\n" // Replace it with newline. This is useful when we have a line with: "Example:\n\tJSON-someting...".
+			} else {
+				line += " "
+			}
+			buffer.WriteString(line)
+		}
+	}
+
+	postDoc := strings.TrimRight(buffer.String(), "\n")
+	postDoc = strings.Replace(postDoc, "\\\"", "\"", -1) // replace user's \" to ".
+	postDoc = strings.Replace(postDoc, "\"", "\\\"", -1) // Escape ".
+	postDoc = strings.Replace(postDoc, "\n", "\\n", -1)
+	postDoc = strings.Replace(postDoc, "\t", "\\t", -1)
+
+	return postDoc
+}
+
+// fieldName extracts the name of the field as it should appear in YAML format and returns the resultant string.
+// "-" indicates that this field is not part of the YAML representation and is thus excluded.
+func fieldName(field *ast.Field) string {
+	const tag = "yaml"
```

Comment:
I am using the [`jsonToYaml`](https://github.com/kubernetes/test-infra/pull/13637/commits/b172776feb114852af5bed163cd2866bd6485f31#diff-a55961ed7ceed93fe28ff3b9d47a8802R127-R142) logic (similar to kubernetes-sigs/yaml) which allows us to leverage the `json` tags. 

Unfortunately, now this ignores `yaml` tags in the marshaling (so it can't be used as a fallback), but since we do not use `yaml` tags we should be 👌here. I can work on adding support for this in the near future.